File: fixtures/scale/java/large/src/cryptobench/ciphers/affine/AffineCipher.java
package cryptobench.ciphers.affine;

import cryptobench.core.Cipher;
import cryptobench.util.Alphabet;

/** Maps x to (a*x + b) mod 26; a must be coprime with 26. */
public class AffineCipher implements Cipher {
    private final AffineKey key;

    public AffineCipher(AffineKey key) {
        this.key = key;
    }

    @Override
    public String name() {
        return "affine";
    }

    @Override
    public String encrypt(String plaintext) {
        StringBuilder sb = new StringBuilder();
        for (char c : Alphabet.clean(plaintext).toCharArray()) {
            sb.append(Alphabet.charAt(key.getA() * Alphabet.indexOf(c) + key.getB()));
        }
        return sb.toString();
    }

    @Override
    public String decrypt(String ciphertext) {
        int inverse = key.inverseOfA();
        StringBuilder sb = new StringBuilder();
        for (char c : Alphabet.clean(ciphertext).toCharArray()) {
            sb.append(Alphabet.charAt(inverse * (Alphabet.indexOf(c) - key.getB())));
        }
        return sb.toString();
    }
}

File: fixtures/scale/dart/large/lib/ciphers/affine/affine_cipher.dart
import '../../core/cipher.dart';
import '../../util/alphabet.dart';
import 'affine_key.dart';

/// Maps x to (a*x + b) mod 26; a must be coprime with 26.
class AffineCipher implements Cipher {
  final AffineKey key;

  AffineCipher(this.key);

  @override
  String name() => 'affine';

  @override
  String encrypt(String plaintext) {
    final sb = StringBuffer();
    for (final c in cleanAlphabet(plaintext).split('')) {
      sb.write(charAt(key.a * indexOf(c) + key.b));
    }
    return sb.toString();
  }

  @override
  String decrypt(String ciphertext) {
    final inverse = key.inverseOfA();
    final sb = StringBuffer();
    for (final c in cleanAlphabet(ciphertext).split('')) {
      sb.write(charAt(inverse * (indexOf(c) - key.b)));
    }
    return sb.toString();
  }
}
